### PR TITLE
chore: bump golang and dependencies

### PR DESCRIPTION
**Pull request by qweeah**
**What this PR does / why we need it**:
This PR bumps Golang to 1.23.4 and upgrades all related modules.

**Please check the following list**:
- [x]  Does the affected code have corresponding tests, e.g. unit test, E2E test?
- [ ]  Does this change require a documentation update?
- [ ]  Does this introduce breaking changes that would require an announcement or bumping the major version?
- [ ]  Do all new files have an appropriate license header?

### CODE REVIEW

File: go.mod
Comment (shizhMSFT) [inline go.mod:1]: Is this file properly upgraded? I do see dependencies are downgraded. Also, do you want to update the dependencies on the e2e test folder?

Comment (qweeah) [reply go.mod:1]: 1. Those files are generated by: 1) Removing `go.sum` file and indirect entries in `go.mod` file and 2) run `go mod tidy`
2. Done the same for test folder and there is nothing to check in expect `go.work.sum`

Comment (shizhMSFT) [reply go.mod:1]: What's your go version? I cannot re-pro with go 1.23.4.

Comment (qweeah) [reply go.mod:1]: Forced changed to main branch modules as discussed offline.

**Review by shizhMSFT — APPROVED** [review]
LGTM

**Pull request merged**